Add eLabFTW file source from file source templates

domgz opened the pull request:
Support user-defined file sources based on `eLabFTWFilesSource`. Add a file source template for eLabFTW and the required documentation.

![Adding eLabFTW as user-defined files source](https://raw.githubusercontent.com/galaxyproject/galaxy/cf60ef4232142ca77dc72069ea0d98d6e6d57fdd/doc/source/admin/user_file_source_form_full_elabftw.png)

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. Run eLabFTW, for example [using Docker Compose](https://github.com/elabftw/elabimg/blob/master/src/docker-compose.yml-EXAMPLE) (or use https://demo.elabftw.net).
  2. Enable the file source template.
  3. Create an account and generate an API Key on eLabFTW (or use the demo account from https://demo.elabftw.net).
  4. Create a new eLabFTW file source on the user preferences page, configuring the endpoint and API Key.
  5. Create an experiment or resource in eLabFTW and attach a file.
  6. Import the file to your history from the eLabFTW file source.
  7. Export your history to the eLabFTW file source.

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

domgz committed:
eLabFTW file source from file source templates

Support user-defined file sources based on `eLabFTWFilesSource`. Add a file source template for eLabFTW and the required documentation.

domgz commented on lib/galaxy/files/sources/elabftw.py:
```diff
@@ -169,14 +177,29 @@ def __init__(self, *args, **kwargs: Unpack[eLabFTWFilesSourceProperties]):
         self._endpoint = kwargs["endpoint"]  # meant to be accessed only from `_get_endpoint()`
         self._api_key = kwargs["api_key"]  # meant to be accessed only from `_create_session()`
 
-    def get_prefix(self) -> Optional[str]:
-        return None
+    def get_prefix(self, user_context: OptionalUserContext = None) -> Optional[str]:
+        endpoint: ParseResult = self._get_endpoint(user_context=user_context)
+        return self.id if self.scheme not in {"elabftw", DEFAULT_SCHEME} else (endpoint.netloc or None)
+        # it would make better sense to return
+        # `self.id if self.scheme == USER_FILE_SOURCES_SCHEME else (endpoint.netloc or None)`, where
+        # `USER_FILE_SOURCES_SCHEME` comes from `galaxy.managers.file_source_instances`; however, that would lead to a
+        # circular import (maybe `USER_FILE_SOURCES_SCHEME` should be moved to a module in a layer deeper than
+        # `galaxy.managers`)
 
     def get_scheme(self) -> str:
-        return "elabftw"
+        return self.scheme if self.scheme and self.scheme != DEFAULT_SCHEME else "elabftw"
+        # it would make better sense to return `self.scheme if self.scheme == USER_FILE_SOURCES_SCHEME else "elabftw"`,
+        # but the same circular import issue as above arises
```
If this ended up being so weird there must be something I am not getting about the scheme... Where are schemes other than "gxfiles" and "gxuserfiles" coming from? The design intended in galaxyproject/galaxy#19319 with a URI that only depends on the configured endpoint is not how things are supposed to work, right?

> File sources reference files via a URI, while eLabFTW uses auto-incrementing positive integers. For more details read galaxyproject/galaxy#18665 [5]. This leads to the need to declare a mapping between said identifiers and Galaxy URIs.
>
> Those take the form ``elabftw://demo.elabftw.net/entity_type/entity_id attachment_id``, where:
> - ``entity_type`` is either 'experiments' or 'resources'
> - ``entity_id`` is the id (an integer in string form) of an experiment or resource
> - ``attachment_id`` is the id (an integer in string form) of an attachment
>
> For the user-defined file sources use case (when users configure one or more instances of the FilesSource via a file source template), Galaxy URIs have a different scheme and authority, taking the form ``gxuserfiles://file_source_id/entity_type/entity_id/attachment_id``, where:
> - ``file_source_id`` is the file source identifier assigned by Galaxy

domgz commented on lib/galaxy/files/sources/elabftw.py:
```diff
@@ -804,7 +849,7 @@ def _write_from(
         entity_type = entity_type.replace("items", "resources")
 
-        return f"elabftw://{location.netloc}/{entity_type}/{entity_id}/{attachment_id}"
+        return f"/{entity_type}/{entity_id}/{attachment_id}"
```
Should've been part of galaxyproject/galaxy#19319 (bugfix).

davelopez replied:
Then, is it better to extract that change in its own PR and target the release_24.2 :+1:

domgz replied:
:+1: [6cdc929](https://github.com/galaxyproject/galaxy/pull/19493/commits/6cdc929716d3779a8e8612cd3ce5ab4f36d6869c)

domgz replied:
> Then, is it better to extract that change in its own PR and target the release_24.2 👍

galaxyproject/galaxy#19541, not targeting 24.2 because eLabFTW is not part of this version.

domgz committed:
Set `writable: true` in file source template for eLabFTW

bgruening commented on lib/galaxy/files/templates/examples/production_elabftw.yaml:
```diff
@@ -0,0 +1,27 @@
+    type: elabftw
+    endpoint: "{{ variables.endpoint }}"
+    api_key: "{{ secrets.api_key }}"
+    writable: true
```
Should we make it an option? Maybe for some use-cases, you don't want to accidentally write into it?

domgz replied:
It may be confusing that people are asked twice if they want to write into it, because the eLabFTW UI already asks you this when generating the key

![grafik](https://github.com/user-attachments/assets/83fed7b9-94e3-4e34-ae67-ae3d64bef38b)

however, that's still problematic because if you use a read-only key, they the file source is listed as writable and that's also confusing.

In other words, we'gotta pick our poison, so indeed we can make it an option.

domgz replied:
The "good solution" is retrieving the API keys via the API

![grafik](https://github.com/user-attachments/assets/005a0b6b-b794-4edb-addb-9ac99fe158fb)

and compare the hash to know if it's read-only or not. Using that information let the file source adjust its `writable` property itself. But that may not be worth the effort, because it involves digging into eLabFTW's code to find out what the hashing algorithm is.

domgz replied:
Done in [c4693d4](https://github.com/galaxyproject/galaxy/pull/19493/commits/c4693d42448552ea1e7ae581f8240c4884f4d350)

davelopez replied:
Thank you @kysrpex, I think providing the option and informing the users in the template description as you have done here is the way to go :+1:

domgz committed:
Make `writable` an option

Let users choose whether Galaxy should be allowed to export data to eLabFTW or they only need to import data to Galaxy.

davelopez reviewed (approved):
Thank you @kysrpex! I tested it and it works as expected now.
